### PR TITLE
(CONT-339) Add top scope facts check

### PR DESCRIPTION
**Pull request by chelnak**
This commit introduces the top_scope_facts check.

The plugin originated here:

https://github.com/mmckinst/puppet-lint-top_scope_facts-check

and was authored by @mmckinst and @seanmil.

Permission has been given to migrate this check to the puppet-lint default plugin set in this PR:

https://github.com/mmckinst/puppet-lint-top_scope_facts-check/pull/11

### CODE REVIEW

File: puppet-lint.gemspec [hunk 1]
```diff
@@ -32,5 +32,5 @@ Gem::Specification.new do |spec|
   ]
   spec.license = 'MIT'
 
-  spec.required_ruby_version = Gem::Requirement.new('>= 2.7'.freeze)
+  spec.required_ruby_version = Gem::Requirement.new('>= 2.5'.freeze)
```
Comment (ekohl) [inline puppet-lint.gemspec:35]: Should this be in a separate PR, if we do decide to do this?

Comment (chelnak) [reply puppet-lint.gemspec:35]: It would enhance visibility I guess.

The reason I included it here was because I kind of see it as part of this change.

Happy either way though! :-)

**Review by GSPatton — APPROVED** [review]
tested locally and works as expected.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -14,10 +14,29 @@ on:
...
+        ruby_version:
+          - '2.5'
+          - '2.7'
+    name: "acceptance (ruby ${{ matrix.ruby_version }})"
     needs: "spec"
     uses: "puppetlabs/cat-github-actions/.github/workflows/gem_acceptance.yml@main"
     secrets: "inherit"
+    with:
+      ruby_version: ${{ matrix.ruby_version }}
+
```
Comment (ekohl) [inline .github/workflows/ci.yml:42]:
```suggestion
```

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -14,10 +14,29 @@ on:
 jobs:
 
   spec:
+    strategy:
+      fail-fast: false
+      matrix:
+        ruby_version:
+          - '2.5'
+          - '2.7'
```
Comment (ekohl) [inline .github/workflows/ci.yml:22]: Perhaps out of scope, but any plans to test on Ruby 3.0 & 3.1? 3.1 is what I use, so I'd expect it to work but it'd be nice to verify it.

For the changelog alone I would submit it as a separate PR.

**Review by ekohl — COMMENTED** [review]
Should the commit also credit the other contributor?

**Comment by chelnak** [comment]
Yes good point. I can get that updated.

With regards to your other comments:
-  I'm leaning towards moving the min ruby change out of this PR

- If we do the above then I can also move the CI changes out too

- Ruby 3.x is on our radar. I think the team will be jumping on this and puppet 8 testing very soon.

File: puppet-lint.gemspec (hunk 1 above)
Comment (bastelfreak) [reply puppet-lint.gemspec:35]: I don't think this downgrade should be in this PR

File: lib/puppet-lint/plugins/top_scope_facts/top_scope_facts.rb [hunk 4]
```diff
@@ -0,0 +1,35 @@
+# Public: A puppet-lint plugin that will check for the use of top scope facts.
+# For example, the fact `$facts['operatingsystem']` should be used over
```
Comment (bastelfreak) [inline lib/puppet-lint/plugins/top_scope_facts/top_scope_facts.rb:2]: I would use a different fact as example, because that's a legacy fact and the legacy-fact-plugin will replace it with `$facts['os']['name']`

File: lib/puppet-lint/plugins/top_scope_facts/top_scope_facts.rb [hunk 5]
```diff
@@ -0,0 +1,35 @@
+# Public: A puppet-lint plugin that will check for the use of top scope facts.
```
Comment (bastelfreak) [inline lib/puppet-lint/plugins/top_scope_facts/top_scope_facts.rb:1]: should the file contain a link to the original author and repo?

File: puppet-lint.gemspec (hunk 1 above)
Comment (chelnak) [reply puppet-lint.gemspec:35]: See https://github.com/puppetlabs/puppet-lint/pull/85#issuecomment-1439164696

Gonna fix this up shortly

File: lib/puppet-lint/plugins/top_scope_facts/top_scope_facts.rb (hunk 5 above)
Comment (chelnak) [reply lib/puppet-lint/plugins/top_scope_facts/top_scope_facts.rb:1]: Resolved

File: lib/puppet-lint/plugins/top_scope_facts/top_scope_facts.rb (hunk 4 above)
Comment (chelnak) [reply lib/puppet-lint/plugins/top_scope_facts/top_scope_facts.rb]: @bastelfreak updated. Happy to use a completely different fact if you think there is a better example.

File: lib/puppet-lint/plugins/top_scope_facts/top_scope_facts.rb [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+# For example, the fact `$facts['os']['name']` should be used over
+# `$::operatingsystem`.
```
Comment (bastelfreak) [inline lib/puppet-lint/plugins/top_scope_facts/top_scope_facts.rb:3]: I still think this is a bad example. `$::operatingsystem` will be converted by this plugin to `$facts['operatingsystem']` and then puppet-lint-legacy_facts-check will convert it into `$facts['os']['name']`

```suggestion
# For example, the fact `$facts['kernel'] should be used over
# `$::kernel`.
```

Comment (chelnak) [reply lib/puppet-lint/plugins/top_scope_facts/top_scope_facts.rb]: Nice one. Updated.

**Review by jordanbreen28 — APPROVED** [review]

**Review by bastelfreak — APPROVED** [review]

**Comment by alexjfisher** [comment]
> Should the commit also credit the other contributor?

But not me? ;)

**Comment by chelnak** [comment]
Oh no! Sorry @alexjfisher. My bad. I can get this updated for you today.

**Comment by alexjfisher** [comment]
> Oh no! Sorry @alexjfisher. My bad. I can get this updated for you today.

Nah, don't worry about it.  It's in the merged commit message so what were you planning on updating anyway? :)

**Comment by chelnak** [comment]
#90 Adds you to the checks description 👍
